usrp_drivers/n200/n200_driver.cpp

```diff
@@ -24,271 +24,339 @@
+    {
+      //TODO(keith): handle error
+    }
 
-  while (1) {
-    packet_socket.recv(&request);
-    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
-    std::cout << "Received in TRANSMIT\n";
-    driverpacket::DriverPacket dp;
-    std::string msg_str(static_cast<char*>(request.data()), request.size());
-    dp.ParseFromString(msg_str);
-
-    sqn_num = dp.sequence_num();
-
+    sqn_num = driver_packet.sequence_num(); // REVIEW #0 default value for numeric types in protobuf is =0 this could create errors, should start at 1?
+                                            // I don't see this being an issue since the only time this is valid is the first pulse seq. If its not incremented on next then the error is caught
     if (sqn_num != expected_sqn_num){
-      std::cout << "SEQUENCE NUMBER MISMATCH: SQN " << sqn_num << " EXPECTED: "
-        << expected_sqn_num << std::endl;
+      DEBUG_MSG("SEQUENCE NUMBER MISMATCH: SQN " << sqn_num << " EXPECTED: " << expected_sqn_num);
       //TODO(keith): handle error
     }
 
-    std::cout <<"TRANSMIT burst flags: SOB "  << dp.sob() << " EOB " << dp.eob() <<std::endl;
-    std::chrono::steady_clock::time_point stream_begin = std::chrono::steady_clock::now();
-    if (dp.channels_size() > 0 && dp.sob() == true && channels_set == false) {
-      std::cout << "TRANSMIT starting something new" <<std::endl;
-      channels = make_channels(dp);
-      stream_args.channels = channels;
-      usrp_d->set_tx_rate(dp.txrate());  // ~450us
-      tx_stream = usrp_d->get_usrp()->get_tx_stream(stream_args);  // ~44ms
-      channels_set = true;
+    DEBUG_MSG("TRANSMIT burst flags: SOB "  << driver_packet.sob() << " EOB "
+      << driver_packet.eob());
+
+    auto stream_setup_begin = std::chrono::steady_clock::now();
+
+    //On start of new sequence, check if there are new USRP channels and if so
+    //set what USRP TX channels and rate(Hz) to use.
+    if (driver_packet.channels_size() > 0 && driver_packet.sob() == true)
+    {
+      DEBUG_MSG("TRANSMIT starting something new");
+      channels = make_tx_channels(driver_packet);
+      stream_args.channels = channels; // REVIEW #15 check that the channels from the driver packet are actually available (a mapping as mentioned in other comment would be handy here)
+                                       // REPLY Need to explain this
+      usrp_d.set_tx_rate(driver_packet.txrate(),channels);  // ~450us REVIEW #15 check that it's not = 0 (default)
+                                                   // REPLY added this check to set_tx_rate
+      tx_stream = usrp_d.get_usrp_tx_stream(stream_args);  // ~44ms REVIEW : idea maybe could we set this up with all channels and then send 0s to any that we don't want to transmit on? then this tx_stream setup could be removed from while loop.
+      usrp_channels_set = true;                                   // REPLY Yes. We will have to see what that looks like on the scope.
```

Comment:
You mean look at the output waveform on the scope to make sure that it's actually 0 and not transmitting anything? Yes.